qemu/tests/iperf_test.py

```diff
@@ -0,0 +1,176 @@
+                parallel_cur = len(re.findall(catch_data, data_info))
+                parallel_exp = int(params.get("parallel_num", 0))
+                if not parallel_cur:
+                    test.fail("iperf client not connected to server")
+                elif parallel_exp and parallel_cur != parallel_exp:
+                    test.fail("Number of parallel threads running(%d) is "
+                              "inconsistent with expectations(%d)"
+                              % (parallel_cur, parallel_exp))
+                logging.info("iperf client successfully connected to server")
+
+    def is_iperf_running(name_pattern, session=None):
+        if session:
+            check_iperf_cmd = params["check_iperf_cmd"] % name_pattern
+            status = serial_session.cmd_status(check_iperf_cmd)
+        else:
+            status = process.system("pgrep -f %s" % name_pattern,
+                                    ignore_status=True, verbose=False)
+        return status == 0
+
+    os_type = params["os_type"]
+    login_timeout = int(params.get("login_timeout", 360))
+    fw_stop_cmd = params["fw_stop_cmd"]
+    tmp_dir = params.get("tmp_dir", "/tmp/")
+    iperf_test_duration = int(params["iperf_test_duration"])
+    iperf_deps_dir = data_dir.get_deps_dir("iperf")
+    host_iperf_file = params["host_iperf_file"]
+    guest_iperf_file = params.get('guest_iperf_file', host_iperf_file)
+    host_src_path = os.path.join(iperf_deps_dir, host_iperf_file)
+    guest_src_path = os.path.join(iperf_deps_dir, guest_iperf_file)
+    host_dst_path = tmp_dir + host_iperf_file
+    guest_dst_path = params.get('guest_dst_path', host_dst_path)
+
+    vm = env.get_vm(params["main_vm"])
+    vm.verify_alive()
+    serial_session = vm.wait_for_serial_login(timeout=login_timeout)
+    guest_session = vm.wait_for_login(timeout=login_timeout)
+    guest_session.cmd(fw_stop_cmd, ignore_all_errors=True)
+
+    vm.copy_files_to(guest_src_path, guest_dst_path)
+    host_ip_addr = utils_net.get_host_ip_address(params)
+    guest_ip_addr = vm.get_address()
+    host_bin_path = iperf_compile(host_src_path, tmp_dir)
+
+    if os_type == 'linux':
+        if not utils_package.package_install("gcc-c++", guest_session):
+            test.cancel("Please install gcc-c++ to proceed")
+        guest_bin_path = iperf_compile(guest_dst_path, tmp_dir, guest_session)
+    else:
+        guest_bin_path = guest_dst_path + guest_iperf_file
+        iperf_deplist = params.get("iperf_deplist")
+        if iperf_deplist:
+            for d_name in iperf_deplist.split(','):
+                dep_path = os.path.join(data_dir.get_deps_dir("iperf"), d_name)
+                vm.copy_files_to(dep_path, guest_dst_path)
+
+    search_pattern = {'host': host_bin_path.replace('src/', 'src/.*'),
+                      'linux': guest_bin_path.replace('src/', 'src/.*'),
+                      'windows': guest_iperf_file}
+
+    if params.get("iperf_server") == params["main_vm"]:
```

Comment:
line 129 ~ 162 does not looking promising, maybe you can think a better solution.
As long as it works, I can accept current implementation.